libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,236 @@
+         retry_connection();
+      });
+   }
+
+   void amqp_error(const char* message) override {
+      wlog("AMQP connection ${a} publishing to \"${e}\" failed: ${m}; retrying", ("a", (std::string)*amqp_address)("m", message)("e", exchange));
+      retry_connection();
+   }
+
+   void pump_queue() {
+      if(stopping || in_flight || !connected || message_deque.empty())
+         return;
+
+      constexpr size_t max_msg_single_transaction = 16;
+      const unsigned msgs_this_transaction = std::min(message_deque.size(), max_msg_single_transaction);
+
+      channel->startTransaction();
+      std::for_each_n(message_deque.begin(), msgs_this_transaction, [this](const std::vector<char>& msg) {
+         channel->publish(exchange, "", msg.data(), msg.size());
+      });
+
+      in_flight = msgs_this_transaction;
+
+      channel->commitTransaction().onSuccess([this](){
+         message_deque.erase(message_deque.begin(), message_deque.begin()+in_flight);
+         logged_exceeded_max_depth = false;
+      })
+      .onFinalize([this]() {
+         in_flight = 0;
+         //unfortuately we don't know if an error is due to something recoverable or if an error is due
+         // to something unrecoverable. To know that, we need to pump the event queue some so that Channel's or
+         // Connection's onError is delivered. Failure to pump the event queue here can result in runaway recursion.
+         ctx.post([this]() {
+            pump_queue();
+         });
+      });
+   }
+
+   std::unique_ptr<AMQP::Address> amqp_address;
+   std::unique_ptr<reliable_amqp_publisher_handler> handler;
+   std::unique_ptr<AMQP::TcpConnection> connection;
+   std::unique_ptr<AMQP::TcpChannel> channel;
+
+   bool connected = false;
+   bool stopping = false;
+
+   unsigned in_flight = 0;
+   std::deque<std::vector<char>> message_deque;
+
+   std::thread thread;
+   boost::asio::io_context ctx;
+   boost::asio::deadline_timer retry_timer{ctx};
+   unsigned next_retry_time = 1;
+
+   const boost::filesystem::path data_file_path;
+
+   const std::string exchange;
+   bool logged_exceeded_max_depth = false;
+};
+
+reliable_amqp_publisher::reliable_amqp_publisher(const std::string& url, const std::string& exchange, const boost::filesystem::path& unconfirmed_path) :
+   my(new reliable_amqp_publisher_impl(url, exchange, unconfirmed_path)) {}
+
+void reliable_amqp_publisher::publish_message(std::vector<char>&& data) {
+   my->ctx.dispatch([this, d=std::move(data)]() mutable {
+      constexpr unsigned max_queued_messages = 1u<<20u;
+      if(my->message_deque.size() > max_queued_messages) {
+         if(my->logged_exceeded_max_depth == false)
+            elog("AMQP connection ${a} publishing to \"${e}\" has reached ${max} unconfirmed messages; further messages will be dropped",
+                 ("a", (std::string)*my->amqp_address)("e", my->exchange)("max", max_queued_messages));
+         my->logged_exceeded_max_depth = true;
```

Comment:
I could see wiring in an `on_max_depth_error` `std::function` which would allow a user of this to do something different; `app().quit()` for example. We don't necessarily need to over engineer this now, so that could be added later if needed.

Reply:
This is a tough one. I can see it argued a number different ways :woman_shrugging: In general, I feel log spam is a bad thing because it'll make it harder to see the trees through the forest. Since this is actually at the error level (not many logs in EOSIO are), it should _really_ stick out in some sort of monitoring solution I hope. Though maybe amqp logs should be on a separate logger so a monitoring system can flag an alarm on _any_ warns & errors from that logger -- really you should never see neither warns nor errors from this class on a healthy running infra.

But yeah, exiting is also a worthy option too.